ledger-core/virtual/statemachine/conveyor_dispatcher_test.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package statemachine
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/stretchr/testify/require"
+
+	"github.com/insolar/assured-ledger/ledger-core/conveyor"
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/dispatcher"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
+)
+
+func handleFactory(_ pulse.Number, _ pulse.Range, input conveyor.InputEvent) (pulse.Number, smachine.CreateFunc) {
+	switch input.(*DispatcherMessage).MessageMeta.Get("error_type") {
+	case "error":
+		return 0, nil
+	default:
+		panic(errors.New("panic"))
+	}
+}
+
+var (
+	machineConfig = smachine.SlotMachineConfig{
+		PollingPeriod:   500 * time.Millisecond,
+		PollingTruncate: 1 * time.Millisecond,
+		SlotPageSize:    1000,
+		ScanCountLimit:  100000,
+	}
+)
+
+func newDispatcherWithConveyor() dispatcher.Dispatcher {
+	ctx := context.Background()
+	pulseConveyor := conveyor.NewPulseConveyor(ctx, conveyor.PulseConveyorConfig{
+		ConveyorMachineConfig: machineConfig,
+		SlotMachineConfig:     machineConfig,
+		EventlessSleep:        0,
+		MinCachePulseAge:      100,
+		MaxPastPulseAge:       1000,
+	}, handleFactory, nil)
+	return NewConveyorDispatcher(pulseConveyor)
+}
+
+func TestConveyorDispatcher_ErrorUnmarshalHandling(t *testing.T) {
+	dispatcher := newDispatcherWithConveyor()
+	require.NotPanics(t, func() {
+		require.NoError(t, dispatcher.Process(&message.Message{}))
+	})
+}
+
+func TestConveyorDispatcher_WrongMetaTypeHandling(t *testing.T) {
+	dispatcher := newDispatcherWithConveyor()
+	req := payload.VCallRequest{}
+	pl, _ := req.Marshal()
+	require.NotPanics(t, func() {
+		require.NoError(t, dispatcher.Process(&message.Message{Payload: pl}))
+	})
+}
+
+func TestConveyorDispatcher_PanicInAddInputHandling(t *testing.T) {
+	dispatcher := newDispatcherWithConveyor()
+	meta := payload.Meta{Pulse: pulse.Number(pulse.MinTimePulse + 1)}
+	metaPl, _ := meta.Marshal()
+	require.NotPanics(t, func() {
+		require.NoError(t, dispatcher.Process(&message.Message{Payload: metaPl}))
+	})
+}
+
+func TestConveyorDispatcher_ErrorInAddInputHandling(t *testing.T) {
+	dispatcher := newDispatcherWithConveyor()
+	meta := payload.Meta{Pulse: pulse.Number(pulse.MinTimePulse + 1)}
+	metaPl, _ := meta.Marshal()
+	metadata := message.Metadata{}
+	metadata.Set("error_type", "error")
+	require.NotPanics(t, func() {
+		require.NoError(t, dispatcher.Process(&message.Message{Payload: metaPl, Metadata: metadata}))
+	})
+}
```

Comment:
I'd suggest adding check of `dispatcher.process` that panics and `dispatcher.Process` that not; otherwise it looks like test checking nothing at all

Reply:
I wrote 2 test:
`TestConveyorDispatcher_PanicInAddInputHandling` - cover case when `handleFactory` raise panic
`TestConveyorDispatcher_ErrorInAddInputHandling` - cover case when `handleFactory` return nil fabric function and other code (`ledger-core/conveyor/conveyor.go#L153`) will return error to `Process`